pgjdbc/src/main/java/org/postgresql/jdbc/PgConnectionFinalizeAction.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2023, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.jdbc;
+
+import org.postgresql.Driver;
+import org.postgresql.util.GT;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Timer;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * This class segregates the minimal resources required for proper cleanup in case
+ * the connection has not been closed by the user code.
+ * <p>For now, it has two actions:</p>
+ * <ul>
+ *   <li>Print stacktrace when the connection has been created, so users can identify the leak</li>
+ *   <li>Release shared timer registration</li>
+ * </ul>
+ */
+class PgConnectionFinalizeAction implements Closeable {
+  private static final Logger LOGGER = Logger.getLogger(PgConnection.class.getName());
+
+  private final ResourceLock lock;
+
+  private @Nullable Throwable openStackTrace;
+  private final Closeable queryExecutorCloseAction;
+
+  // Timer for scheduling TimerTasks for the connection.
+  // Only instantiated if a task is actually scheduled.
+  private volatile @Nullable Timer cancelTimer;
```

Comment:
does this really need to be volatile? It appears to only ever be accessed while holding the ResourceLock.